lib/puppet/state_machine/machine_runner.rb

```diff
@@ -0,0 +1,46 @@
+require 'puppet/state_machine'
+
+module Puppet
+  class StateMachine
+    # Run a state machine until it enters a final or error state.
+    #
+    # This runner follows the behavior of a DFA in that the machine is run until completion or
+    # error. There are two notable differences from this runner and a pure DFA. First, there
+    # is no input string, and instead the state event callbacks determine the next transition.
+    # Second, instead of terminating the state machine successfully if the machine is in a final
+    # state when all input is consumed, entering a final state will immediately terminate the
+    # machine.
+    class MachineRunner
```

Comment:
I'm really enjoying how many comments GitHub is making me rewrite because of how intuitive and well behaved the review functionality works. In any case my immediate need is for a state machine that behaves like a DFA in that it runs to completion, but that's not a required property of a state machine. State machines could be used to model the [Solaris zone state machine](https://github.com/puppetlabs/puppet/blob/master/lib/puppet/type/zone.rb#L10). Outside operations transition the state machine between states, and the machine itself doesn't run till completion - in fact it doesn't run at all. This machine running logic is only needed for a certain class of machines and the state machine implementation as it stands doesn't have to be constrained to that single use case.